pkg/plugins/runtime/gateway/filter_chain_generator.go

```diff
@@ -437,11 +364,123 @@ func (g *TCPFilterChainGenerator) Generate(
 	)
 
 	if info.Listener.Protocol == mesh_proto.MeshGateway_Listener_TLS {
+		tlsResources, err := configureTLS(
+			ctx,
+			info,
+			info.HostInfos[0].Host.TLS,
+			sniNames,
+			builder,
+			nil,
+		)
+		resources = resources.AddSet(tlsResources)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	return resources, []*envoy_listeners.FilterChainBuilder{builder}, nil
+}
+
+func configureTLS(
+	ctx xds_context.Context,
+	listener GatewayListenerInfo,
+	tls *mesh_proto.MeshGateway_TLS_Conf,
+	hostnames []string,
+	builder *envoy_listeners.FilterChainBuilder,
+	alpnProtocols []string,
+) (
+	*core_xds.ResourceSet, error,
+) {
+	builder.Configure(
+		envoy_listeners.MatchTransportProtocol("tls"),
+	)
+
+	resources := core_xds.NewResourceSet()
+
+	downstream := newDownstreamTypedConfig(alpnProtocols)
+
+	mode := mesh_proto.MeshGateway_TLS_NONE
+	if tls != nil {
+		mode = tls.GetMode()
+	}
+
+	switch mode {
+	case mesh_proto.MeshGateway_TLS_PASSTHROUGH:
+		builder.Configure(
+			envoy_listeners.MatchServerNames(hostnames...),
+		)
+		return resources, nil
+	case mesh_proto.MeshGateway_TLS_TERMINATE:
 		builder.Configure(
-			envoy_listeners.MatchTransportProtocol("tls"),
-			envoy_listeners.MatchServerNames(sniNames...),
+			envoy_listeners.MatchServerNames(hostnames...),
+		)
+
+		// Note that Envoy 1.184 and earlier will only accept 1 SDS reference.
```

Comment:
we don't care about 1.18.4 anymore right? The support for it is long gone I think